src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -3101,5 +3101,218 @@ internal virtual ImmutableArray<LabelSymbol> Labels
+
+                // SPEC VIOLATION: The spec says we should apply overload resolution, but Dev10 uses
+                // some custom logic in ExpressionBinder.BindGrpToParams.  The biggest difference
+                // we've found (so far) is that it only considers methods with zero parameters
+                // (i.e. doesn't work with "params" or optional parameters).
+                if (!method.Parameters.Any())
+                {
+                    candidateMethods.Add((MethodSymbol)member);
+                }
+            }
+            MethodSymbol patternMethod = PerformPatternOverloadResolution(patternType, candidateMethods, syntaxExpr, warningsOnly, diagnostics, syntaxTree, messageID);
+
+            candidateMethods.Free();
+
+            return patternMethod;
+        }
+
+        /// <summary>
+        /// The overload resolution portion of FindForEachPatternMethod.
+        /// </summary>
+        private MethodSymbol PerformPatternOverloadResolution
+            (TypeSymbol patternType, ArrayBuilder<MethodSymbol> candidateMethods,
+            SyntaxNode syntaxExpression, bool warningsOnly, DiagnosticBag diagnostics,
+            SyntaxTree syntaxTree, MessageID messageID)
+        {
+            ArrayBuilder<TypeSymbol> typeArguments = ArrayBuilder<TypeSymbol>.GetInstance();
+            AnalyzedArguments arguments = AnalyzedArguments.GetInstance();
+            OverloadResolutionResult<MethodSymbol> overloadResolutionResult = OverloadResolutionResult<MethodSymbol>.GetInstance();
+
+            HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+            // We create a dummy receiver of the invocation so MethodInvocationOverloadResolution knows it was invoked from an instance, not a type
+            var dummyReceiver = new BoundImplicitReceiver(syntaxExpression, patternType);
+            this.OverloadResolution.MethodInvocationOverloadResolution(
+                methods: candidateMethods,
+                typeArguments: typeArguments,
+                receiver: dummyReceiver,
+                arguments: arguments,
+                result: overloadResolutionResult,
+                useSiteDiagnostics: ref useSiteDiagnostics);
+            diagnostics.Add(syntaxExpression, useSiteDiagnostics);
+
+            MethodSymbol result = null;
+
+            if (overloadResolutionResult.Succeeded)
+            {
+                result = overloadResolutionResult.ValidResult.Member;
+
+                if (result.IsStatic || result.DeclaredAccessibility != Accessibility.Public)
+                {
+                    if (warningsOnly)
+                    {
+                        diagnostics.Add(ErrorCode.WRN_PatternStaticOrInaccessible, syntaxExpression.Location, patternType, messageID.Localize(), result);
+                    }
+                    result = null;
+                }
+                else if (messageID == MessageID.IDS_Disposable && !result.ReturnsVoid && warningsOnly)
```

Comment:
How is this `if` statement different from the above? It seems like more of this code could be shared and it would prevent us fram having to switch on the `messageID`, which is a bit of a code smell. Seems like either this code is similar enough that we could get more sharing, or it's different enough that we might as well give up on sharing and write two methods.

Reply:
Will move these to the individual `Satisfies___Pattern` methods in UsingStatementBinder.cs and ForEachLoopBinder.cs